lib/Optimizer/Transforms/LiftArrayAlloc.cpp

```diff
@@ -249,11 +286,19 @@ class AllocaPattern : public OpRewritePattern<cudaq::cc::AllocaOp> {
               scoreboard[0] = w;
               continue;
             }
-          return false;
+          // can be a cast only used for a quake.init_state or vector init
+          continue;
+        } else {
+          if (getWriteOp(cast, 0)) {
+            LLVM_DEBUG(llvm::dbgs()
+                       << "another cast used in store: " << *op << '\n');
+            return false;
+          }
+          // can be a cast only used for a quake.init_state or vector init
+          toGlobal = true;
+          continue;
```

Comment:
These changes look suspect to me. I assume they are to workaround the problems of floating-point conversions being kicked down the road and having floating point extensions and truncations being inserted?

There is only 1 truly safe cast that should be allowed. Specifically, casting the array of element types pointer to a singleton of element type for the single intention of writing a value of element type to that position in memory. Casting to arbitrary types other than the element type or having random other operations on the casting of the array's address would open the door to bugs.

Reply:
No transformations should be moved from their original places, and I added a removal of ext and trunc so it is done before we get here. I just had more test cases coming from my tests and made sure they all worked (ie globals coming from ConvertExpr, LiftArray, QuakeSynth, as well as the test you added for unitary synthesis. Those changes make all of those test cases work. This specific line is only needed for the vector init, which is the unitary synthesis case.

Reply:
Examples of an extra cast:

Unitary synthesis:
```
%9 = cc.cast %4 : (!cc.ptr<!cc.array<complex<f64> x 4>>) -> !cc.ptr<i8>
%10 = call @__nvqpp_vectorCopyCtor(%9, %c4_i64, %c16_i64) : (!cc.ptr<i8>, i64, i64) -> !cc.ptr<i8>
%11 = cc.stdvec_init %10, %c4_i64 : (!cc.ptr<i8>, i64) -> !cc.stdvec<complex<f64>>
```

Reply:
Reverted the changes and added a separate pass that removes the extra cast.